### PR TITLE
dotnet update package for NuGet packages.

### PR DESCRIPTION
**Pull request by JonDouglas**
Introduction to an updating experience in the .NET CLI that allows one to update dependencies in projects and solutions while supporting helpful options to stay current based on an individuals update strategy.

This proposal introduces one new command to .NET CLI known as `dotnet update package`.

[Rendered Spec](https://github.com/NuGet/Home/blob/4ea15dd4f7f9431310e71432e8a31555d1b69517/proposed/2022/DotnetUpdatePackage.md)

Please 👍 or 👎 this comment to help us with the direction of this feature & leave as much feedback/questions/concerns as you'd like on this issue itself and we will get back to you shortly.

Thank You 🎉

### CODE REVIEW

File: proposed/2022/DotnetUpdatePackage.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+dotnet update package Microsoft.*
+```
+
+- Displays what would be updated in the update operation for the current working directory.
+
+```
+dotnet update package --dry-run
+```
+
+#### Exit Codes
```
Comment (chrisraygill) [inline proposed/2022/DotnetUpdatePackage.md:127]: Is it all or nothing? Is it possible for some packages to update while others fail?

File: proposed/2022/DotnetUpdatePackage.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+- [dart pub upgrade](https://dart.dev/tools/pub/cmd/pub-upgrade)
+- [pip install -U](https://pip.pypa.io/en/stable/cli/pip_install/)
+
+## Unresolved Questions
+
+<!-- What parts of the proposal do you expect to resolve before this gets accepted? -->
+<!-- What parts of the proposal need to be resolved before the proposal is stabilized? -->
+<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->
+- Should this command use `dotnet add package` under the hood for the scenarios that make sense such as finding the latest version of a specific package name?
+- What verb makes most sense? `update`, `upgrade`, `refresh`, etc? Update seems the most consistent for package managers such as cargo, npm, rubygems, and NuGet. Upgrade seems the most consistent for package managers such as pip.
```
Comment (chrisraygill) [inline proposed/2022/DotnetUpdatePackage.md:172]: Worth noting that "update" is consistent with the current PMUI terminology.

File: proposed/2022/DotnetUpdatePackage.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+- --prerelease
+
+Allows prerelease packages to be installed. Available since .NET Core 5 SDK
+
+- -s|--source <SOURCE>
+
+The URI of the NuGet package source to use during the restore operation.
+
+- -v|--version <VERSION>
```
Comment (chrisraygill) [inline proposed/2022/DotnetUpdatePackage.md:85]: Can you specify version for a glob of packages as well?

File: proposed/2022/DotnetUpdatePackage.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+Version of the package. See NuGet package versioning.
+
+- --dry-run
```
Comment (chrisraygill) [inline proposed/2022/DotnetUpdatePackage.md:89]: Will this display the changes in transitive packages as well?

File: proposed/2022/DotnetUpdatePackage.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    [-n|--no-restore] [--package-directory <PACKAGE_DIRECTORY>]
+    [--prerelease] [-s|--source <SOURCE>] [-v|--version <VERSION>]
+    [--dry-run] [--highest-minor] [--highest-patch]
+
+dotnet update package -h|--help
+```
+
+#### Arguments
+
+- PROJECT | SOLUTION
```
Comment (chrisraygill) [inline proposed/2022/DotnetUpdatePackage.md:43]: Is there way to specify a directory of projects? i.e. I want to update all my packages in my \test directory?

**Review by chrisraygill — COMMENTED** [review]
Looks awesome overall! Main thing it's missing is example outputs.

File: proposed/2022/DotnetUpdatePackage.md (hunk 1 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:127]: Still thinking through it. Any suggestions?

File: proposed/2022/DotnetUpdatePackage.md (hunk 3 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:85]: We will need to lock down the update strategies first I think. In general, I don't think so unless people have a compelling reason to do this!

File: proposed/2022/DotnetUpdatePackage.md (hunk 4 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:89]: Top-level for now since those would be the only "affected" packages per-say. But let's keep this open for more perspectives.

File: proposed/2022/DotnetUpdatePackage.md (hunk 5 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:43]: We'll need to limit the scope. So let's start with project or solution. Updating the world would be amazing, but very complex I think!

File: proposed/2022/DotnetUpdatePackage.md (hunk 3 above)
Comment (chrisraygill) [reply proposed/2022/DotnetUpdatePackage.md:85]: I remember having a conversation with @nkolev92 about how some packages need to have their version kept in sync, particularly in ASP.NET Core world. I believe a lot of those packages also share prefixes like these:

image

Azure packages are also versioned similarly:

image

In these cases `dotnet projectA update package Azure.Storage.* 12.10.0` might be pretty handy

File: proposed/2022/DotnetUpdatePackage.md (hunk 4 above)
Comment (chrisraygill) [reply proposed/2022/DotnetUpdatePackage.md:89]: IMO, top-level only by default but have a --verbosity option to include more info.

File: proposed/2022/DotnetUpdatePackage.md (hunk 5 above)
Comment (chrisraygill) [reply proposed/2022/DotnetUpdatePackage.md:43]: Is it? This design already necessitates the ability to "find a project" within the current directory when the `<PROJECT>` isn't specified. Might be oversimplifying, but isn't the functionality I'm describing just a loop to do that for all projects recursively within a directory?

Conceivably, one of the biggest users of this feature will be VS Code .NET devs who might not have a solution file to target. However, we know that many users prefer to keep all of their packages across all projects at the latest stable version.

It also adds a nice algorithmic component for an intern 😉

File: proposed/2022/DotnetUpdatePackage.md (hunk 3 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:85]: That should be supported with current spec right?

File: proposed/2022/DotnetUpdatePackage.md (hunk 5 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:43]: I believe it's possible. For example if you ran the command on the repo folder, it should work recursively.

File: proposed/2022/DotnetUpdatePackage.md (hunk 3 above)
Comment (chrisraygill) [reply proposed/2022/DotnetUpdatePackage.md:85]: Yes, unless explicitly disallowed. I just wanted to clarify the intention 🙂

File: proposed/2022/DotnetUpdatePackage.md (hunk 5 above)
Comment (chrisraygill) [reply proposed/2022/DotnetUpdatePackage.md:43]: Based on this line:

> The project or solution file to operate on. If not specified, the command searches the current directory for one. If more than one solution or project is found, an error is thrown.

The current design will error out if you run it on the repo folder because multiple projects will be found. My expectation would be that it would run recursively/ apply to all found projects.

File: proposed/2022/DotnetUpdatePackage.md (hunk 1 above)
Comment (chrisraygill) [reply proposed/2022/DotnetUpdatePackage.md:127]: IMO, it shouldn't be all or nothing by default - but may want to provide an option for it if users want to take advantage of the exit code - but that's a secondary need.

If I have a project with an older framework where the latest version of half my packages still support my TFM, but the latest version of the other half don't, an all or nothing design would prevent me from bulk updating altogether.

Based on feedback about version preferences from the HaTS survey, I think the typical user will likely want to update all compatible packages to the latest stable version, even if other will fail. I think it's probably a fairly small subset that will take advantage of the exit code for automation purposes.

Maybe `--atomic` or `--disable-partial` would make sense as an option for all or nothing 🤷‍♂️

File: proposed/2022/DotnetUpdatePackage.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+- PACKAGE_NAME
+
+The package reference to add. Package name wildcards should be supported and update a subset of a glob-style pattern package name. i.e. `Microsoft.*` would update Microsoft packages. The commonly used symbols to support are:
+
+- `*` - zero or more characters.
+- `?` - single occurrence of any character.
+- `.` - literal "." character
+
+#### Options
```
Comment (chrisraygill) [inline proposed/2022/DotnetUpdatePackage.md:55]: Any plan to include a `--compatible` or have it automatically update packages to the latest compatible version if the package latest stable version doesn't support the TFM?

It's referenced in this comment: https://github.com/NuGet/Home/issues/4103#issuecomment-802598457

It's probably beyond the scope an initial release, but a maybe a cool idea 🤷‍♂️

Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:55]: Not yet. We don't have compat client tooling yet outside of one field. Once we can use the server side stuff, perhaps we can light up this.

File: proposed/2022/DotnetUpdatePackage.md (hunk 5 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:43]: Let's change this language then!

**Comment by Craige** [comment]
> To ensure compatibility with existing package definitions in various files, we should not overwrite or simplify XML elements, but rather only update package version attributes.

To be clear, please do not update the version string in the csproj file. This is what lock files are for. What I set in csproj should be a resolvable version constraint, not an exact pinned version.

File: proposed/2022/DotnetUpdatePackage.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+#### Arguments
+
+- PROJECT | SOLUTION
+
+The project or solution file to operate on. If not specified, the command searches the current directory for one. If more than one solution or project is found, an error is thrown.
+
+- PACKAGE_NAME
+
+The package reference to add. Package name wildcards should be supported and update a subset of a glob-style pattern package name. i.e. `Microsoft.*` would update Microsoft packages. The commonly used symbols to support are:
```
Comment (nkolev92) [inline proposed/2022/DotnetUpdatePackage.md:49]: Do we know if there's an ask for this at this point?
Would a list of package ids be preferable instead?

File: proposed/2022/DotnetUpdatePackage.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+- `*` - zero or more characters.
+- `?` - single occurrence of any character.
+- `.` - literal "." character
+
+#### Options
+
+- -f|--framework <FRAMEWORK>
```
Comment (nkolev92) [inline proposed/2022/DotnetUpdatePackage.md:57]: This becomes tricky quickly in the solution case. 

For example, what if a project in a solution does not target that exact framework?

File: proposed/2022/DotnetUpdatePackage.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+Adds a package reference only when targeting a specific framework.
```
Comment (nkolev92) [inline proposed/2022/DotnetUpdatePackage.md:59]: Probably meant to say update.

File: proposed/2022/DotnetUpdatePackage.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+Allows the command to stop and wait for user input or action. For example, to complete authentication.
+
+- -n|--no-restore
+
+Adds a package reference without performing a restore preview and compatibility check.
+
+- --package-directory <PACKAGE_DIRECTORY>
+
+The directory where to restore the packages. The default package restore location is %userprofile%\.nuget\packages on Windows and ~/.nuget/packages on macOS and Linux. For more information, see Managing the global packages, cache, and temp folders in NuGet.
```
Comment (nkolev92) [inline proposed/2022/DotnetUpdatePackage.md:75]: I wonder if we can not add this.
I know add package has it, but I'm wondering how commonly it's used if at all. 

Specifying the global packages folder makes more scene in commandline restore scenarios that add/update imo.

File: proposed/2022/DotnetUpdatePackage.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+- --prerelease
+
+Allows prerelease packages to be installed. Available since .NET Core 5 SDK
```
Comment (nkolev92) [inline proposed/2022/DotnetUpdatePackage.md:79]: nit: Copied from add package? :D Probably remove the last part.

File: proposed/2022/DotnetUpdatePackage.md (hunk 3 above)
Comment (nkolev92) [reply proposed/2022/DotnetUpdatePackage.md:85]: Would it be preferrable to specify a list of ids instead?

File: proposed/2022/DotnetUpdatePackage.md (hunk 4 above)
Comment (nkolev92) [reply proposed/2022/DotnetUpdatePackage.md:89]: I think we should start with top level ones, but we can consider adding it later. 

It's probably worth considering whether we do deltas or maybe just display the new list of packages.

File: proposed/2022/DotnetUpdatePackage.md [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+The URI of the NuGet package source to use during the restore operation.
+
+- -v|--version <VERSION>
+
+Version of the package. See NuGet package versioning.
+
+- --dry-run
+
+Displays what would be updated, but doesn't actually do the operation.
```
Comment (nkolev92) [inline proposed/2022/DotnetUpdatePackage.md:91]: Default is major? 

Wonder if we need an option that takes parameters instead?

File: proposed/2022/DotnetUpdatePackage.md (hunk 1 above)
Comment (nkolev92) [reply proposed/2022/DotnetUpdatePackage.md:127]: I think `update by tfm compatibility` is a special scenario. 
It's one of those few scenarios where a failure is that exact package only. 

The trickier one is what if the latest versions causes resolution conflicts, such as downgrades etc. 
It becomes a user decision what they'd want to do then, so something semi interactive.  
`--no-restore` is a an option that *skips* validations.
Right now the most common failures at, NU1605, NU1105, NU1201 and NU1104.

Back to the compat scenario, I think it's the one that we're most likely to solve. Worth noting that we'd have performance/implementation challenges for any source that isn't nuget.org.
Right now, search by tfm is a search thing only, we'd need to validate the perf issues. 

tldr; I feel the pain of the customers, I run into some of these myself, but solving this in the first iteration might be a reach.

File: proposed/2022/DotnetUpdatePackage.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+- [RubyGems update](https://rubygems.org/gems/rubygems-update)
+- [dart pub upgrade](https://dart.dev/tools/pub/cmd/pub-upgrade)
+- [pip install -U](https://pip.pypa.io/en/stable/cli/pip_install/)
+
+## Unresolved Questions
+
+<!-- What parts of the proposal do you expect to resolve before this gets accepted? -->
+<!-- What parts of the proposal need to be resolved before the proposal is stabilized? -->
+<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->
+- Should this command use `dotnet add package` under the hood for the scenarios that make sense such as finding the latest version of a specific package name?
```
Comment (nkolev92) [inline proposed/2022/DotnetUpdatePackage.md:171]: I think that's a technical question that'll likely be accounted for by the implementer. 
These 2 commands will share a bunch of common functionality.

File: proposed/2022/DotnetUpdatePackage.md [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+- What verb makes most sense? `update`, `upgrade`, `refresh`, etc? Update seems the most consistent for package managers such as cargo, npm, rubygems, and NuGet. Upgrade seems the most consistent for package managers such as pip.
+- Should `--highest-minor` and `--highest-patch` be replaced with a single `--dependency-version <version>` parameter to account for more scenarios like `Lowest(default)` and `Highest`?
```
Comment (nkolev92) [inline proposed/2022/DotnetUpdatePackage.md:173]: The `Lowest` vs `Highest` is a larger feature than the update command itself. 
it makes it difficult for me to suggest a name based on what that feature may turn out to be.

File: proposed/2022/DotnetUpdatePackage.md (hunk 5 above)
Comment (nkolev92) [reply proposed/2022/DotnetUpdatePackage.md:43]: How many folders deep would you search for projects? 
All of it? 

I think deciding the `target` of a command should be something we involve the SDK team on given that `package` is a first class noun in dotnet.exe.

**Review by nkolev92 — COMMENTED** [review]
This proposals brings up all the discussion points that we'll need to have when we implement this features.

Thanks for the initial write-up @JonDouglas 

Most of my comments are doc readiness, but a good amount are about the functionality itself as well.

File: proposed/2022/DotnetUpdatePackage.md (hunk 11 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:79]: Yes for consistency sake!

File: proposed/2022/DotnetUpdatePackage.md (hunk 7 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:49]: The primary ask is updating all packages. The secondary ask is having control of updating. A list of package IDs can be preferable for sure. This shorthand might be helpful in our ecosystem given prefix/source mapping.

File: proposed/2022/DotnetUpdatePackage.md (hunk 5 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:43]: @baronfel Any idea how we can get more eyes from SDK team on these types of things? :)

File: proposed/2022/DotnetUpdatePackage.md (hunk 10 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:75]: Sadly we don't have much of a signal of how much this is used. Does this still make sense for update scenario?

File: proposed/2022/DotnetUpdatePackage.md (hunk 12 above)
Comment (JonDouglas) [reply proposed/2022/DotnetUpdatePackage.md:91]: Open to perspectives! Right now yes, major default with two flags for minor and patch.

File: proposed/2022/DotnetUpdatePackage.md (hunk 5 above)
Comment (baronfel) [reply proposed/2022/DotnetUpdatePackage.md:43]: You can unfortunately can't tag @dotnet/dotnet-cli, so I'll poke a few here: @dsplaisted, @joeloff, @gkulin

I'd generally be in favor of globs/patterns for more SDK operations - I know a similar capability has been requested for many of the MSBuild-driving commands in the past.  It's common to use globs for this kind of thing - `**/*Test.*proj` to grab every test project in a solution recursively, for example.

File: proposed/2022/DotnetUpdatePackage.md [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+dotnet update package -h|--help
+```
+
+#### Arguments
+
+- PROJECT | SOLUTION
+
+The project or solution file to operate on. If not specified, the command searches the current directory for one. If more than one solution or project is found, an error is thrown.
+
+- PACKAGE_NAME
```
Comment (baronfel) [inline proposed/2022/DotnetUpdatePackage.md:47]: going to throw out a crazy idea here that might consolidate a few flags/options.

`PACKAGE_NAME` can be
* pattern as described below, or
* a list of 'package with version spec's, where 'package with version spec' means a string of the format `{PACKAGE_NAME}@{npm-style version expression}`

The NPM(and also gem/paket/etc)-style expressions allow you to easily express bounds for the update operation on a package-by package basis, unambiguously, in a way that `PACKAGENAME -version VERSION` can't scale to.  It also removes the need for `--higest-minor`, `--highest-patch`, and `--prerelease`, as those can be specified with appropriate syntax in the correct positions.

This does come with a downside of a) documenting these and b) writing code to parse them, however.

File: proposed/2022/DotnetUpdatePackage.md (hunk 1 above)
Comment (OskarKlintrot) [reply proposed/2022/DotnetUpdatePackage.md:127]: Another "fun" case when updating by tfm compatibility that I run in to is that dotnet-ef supports something like dotnet core 2.1 in the latest version so if you update everything to the latest compatible version the EF Core packages might get out-of-sync. I really whish to be to update all packages to the latest compatible version but it's not as easy as it might seem.

**Comment by Frulfump** [comment]
What's the status of this, some of it is implemented I think?

**Comment by zivkan** [comment]
I need to review the PR and see what ideas it proposed that are missing from the already implemented `dotnet package update` command, and create issues to track them.  But this PR will get closed, as https://github.com/NuGet/Home/blob/dev/accepted/2024/package-update-command.md supersedes this PR.

**Comment by Frulfump** [comment]
> This PR has been automatically marked as stale because it has no activity for **30 days**. It will be closed if no further activity occurs **within another 15 days** of this comment, unless it has a "Status:Do not auto close" label. If it is closed, you may reopen it anytime when you're ready again, as long as you don't delete the branch.
> 

See previous comment. Can we get the do not auto close label please?

**Comment by Frulfump** [comment]
Waiting for review and then this PR getting closed as per https://github.com/NuGet/Home/pull/11812#issuecomment-4027439686 but if we let it get closed it will most likely fade into the ether

**Comment by zivkan** [comment]
Oops, I forgot to close this out.  When I reviewed, I/copilot found two features this spec proposed without tracking issues, so I created them:

* https://github.com/NuGet/Home/issues/14870
* https://github.com/NuGet/Home/issues/14871

Copilot said there was at least one other feature this spec proposed that is not implemented in the current `dotnet package update`, but there was already an issue for it. Unfortunately I can't remember what issue that is.

Maybe we should consider adding an `Area:DotnetPackageUpdate` label, to make it easier to find issues specific to it, but for now what I do is [put "dotnet package update" in quotes, and also use the `in:title` filter](https://github.com/NuGet/Home/issues?q=is%3Aissue%20state%3Aopen%20in%3Atitle%20%22dotnet%20package%20update%22)

**Comment by Frulfump** [comment]
> Oops, I forgot to close this out. When I reviewed, I/copilot found two features this spec proposed without tracking issues, so I created them:
> 
>     * [Add `--dry-run` to `dotnet package update` #14870](https://github.com/NuGet/Home/issues/14870)
> 
>     * [Add `--deprecated` option to `dotnet package update` #14871](https://github.com/NuGet/Home/issues/14871)
> 
> 
> Copilot said there was at least one other feature this spec proposed that is not implemented in the current `dotnet package update`, but there was already an issue for it. Unfortunately I can't remember what issue that is.
> 
> Maybe we should consider adding an `Area:DotnetPackageUpdate` label, to make it easier to find issues specific to it, but for now what I do is [put "dotnet package update" in quotes, and also use the `in:title` filter](https://github.com/NuGet/Home/issues?q=is%3Aissue%20state%3Aopen%20in%3Atitle%20%22dotnet%20package%20update%22)

Ah no worries, thanks! Area label sounds good and maybe even an epic to track all of them? Appreciate the search query in the meantime
